index.bs

```diff
@@ -175,6 +175,88 @@ partial interface Navigator {
 };
 </pre>
 
+# Audio session algorithms # {#audio-session-algorithms}
+
+## Update AudioSession's type ## {#audio-session-update-type-algorithm}
+
+To <dfn>update the type</dfn> of |audioSession|, the user agent MUST run the following steps:
+1. If |audioSession|.[=AudioSession/[[isTypeBeingApplied]]=] is `true`, abort these steps.
+1. Set |audioSession|.[=AudioSession/[[isTypeBeingApplied]]=] to `true`.
+1. [=Queue a task=] to run the following steps:
+    1. Set |audioSession|.[=AudioSession/[[isTypeBeingApplied]]=] to `false`.
+    1. If |audioSession|.[=AudioSession/[[type]]=] is the same as |audioSession|.[=AudioSession/[[appliedType]]=], abort these steps.
+    1. Set |audioSession|.[=AudioSession/[[appliedType]]=] to |audioSession|.[=AudioSession/[[type]]=].
+    1. [=Update all AudioSession states=] of |audioSession|'s [=top-level browsing context=] with |audioSession|.
+    1. For each |element| of |audioSession|.[=AudioSession/[[elements]]=], [=update an element|update=] |element|.
+    1. Let |newType| be the result of [=compute the audio session type|computing the type=] of |audioSession|.
+    1. [=In parallel=], change the [=audio session/type=] of |audioSession|'s [=audio session=] to |newType|.
+
+## Update AudioSession's state ## {#audio-session-update-state-algorithm}
+
+When an audio session [=audio session/element=] is starting or stopping, the user agent will run steps that <dfn>change the state</dfn> of an [=audio session=].
```

Comment:
We would probably need another PR to define `start and stop` for different audio session element. We also need to define when this algorithm will really need to be run.

Eg. if a AudioSession has several different audio session elements, only one of them stops, but other are still playing, in this case, should this algorithm be triggered?

Reply:
This is done in https://github.com/w3c/audio-session/pull/34.
See https://pr-preview.s3.amazonaws.com/youennf/audio-session/pull/34.html#integration

> if a AudioSession has several different audio session elements, only one of them stops, but other are still playing, in this case, should this algorithm be triggered?

The UA should run steps to decide whether to inactivate the audio session (and decide to leave the audio session active).
The algorithm in https://pr-preview.s3.amazonaws.com/youennf/audio-session/pull/34.html#integration) needs to be revised.

Reply:
Sounds good, let's do the modification in #34.